### PR TITLE
🩺 Dr. GRPO loss

### LINKED ISSUE #3178: the loss is dapo not the dr.grpo ?
https://github.com/huggingface/trl/blob/main/trl/trainer/grpo_trainer.py#L961

the current loss is $\frac1{\sum_{i=1}^G\|o_1\|}\sum_{i}^G\sum_{t=1}^{\|o_i\|}...$ (dapo-like) not the dr.grpo $\frac1{G}\sum_{i}^G\sum_{t=1}^{\|o_i\|}...$ (https://huggingface.co/docs/trl/main/en/grpo_trainer).

Is my understanding correct?

### PR DESCRIPTION
**Pull request by qgallouedec**
# What does this PR do?

This PR supersedes #3231 #3138
Closes #3178

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -1047,7 +1050,13 @@ def _compute_loss(self, model, inputs):
         per_token_loss = -torch.min(per_token_loss1, per_token_loss2)
         if self.beta != 0.0:
             per_token_loss = per_token_loss + self.beta * per_token_kl
-        loss = (per_token_loss * completion_mask).sum() / completion_mask.sum()
+
+        if self.loss_type == "bnpo":
+            loss = (per_token_loss * completion_mask).sum() / completion_mask.sum()
+        elif self.loss_type == "drgrpo":
+            loss = (per_token_loss * completion_mask).sum() / (per_token_loss.size(0) * self.max_completion_length)
```
Comment (ScottHoang) [inline trl/trainer/grpo_trainer.py:1063]: Hi, Thank you for your contribution. I may be entirely wrong, but shouldn't the denominator here be `self.num_generation`?  Since Dr.GRPO removed 1/|o_i| ?

**Comment by qgallouedec** [comment]
Same effective batch size (256)

| GPUs | Grad accum steps | Per device batch size | 
| -- | ------------------ | -------------------- |
| 1 | 2 | 128 |
| 1 | 4 | 64 |
| 1 | 8 | 32 |
| 1 | 16 | 16 |
| 1 | 32 | 8 |
| 2 | 1 | 128 |
| 2 | 2 | 64 |
| 2 | 4 | 32 |
| 2 | 8 | 16 |
| 2 | 16 | 8 |
| 4 | 1 | 64 |
| 4 | 2 | 32 |
| 4 | 4 | 16 |
| 4 | 8 | 8 |
| 8 | 1 | 32 |
| 8 | 2 | 16 |
| 8 | 4 | 8 |


Screenshot 2025-04-07 at 11 57 02

File: trl/trainer/grpo_trainer.py (hunk 1 above)
Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py:1063]: Yes, as long as it's constant, it doesn't matter.

The motivation behind multiplying by `self.max_completion_length` is that it introduces consistency in terms of loss amplitude for different values of `self.max_completion_length`. Without it, we would probably have to use different learning rates depending on the value of `self.max_completion_length` we choose.

As for `self.num_generations`, it's somehow hidden in `per_token_loss.size(0)`. E.g., in the default configuration, `self.num_generations==per_token_loss.size(0)`. It's important to use `per_token_loss.size(0)` and not `self.num_generations` for the same reason as explained above, it allows you to have consistency in loss amplitude for different batch size values.

Comment (ScottHoang) [reply trl/trainer/grpo_trainer.py:1063]: Thanks for the explanation!

File: docs/source/grpo_trainer.md [hunk 2]
```diff
@@ -92,26 +92,55 @@ $$
...
+$$
+
+
+Furthermore, it was demonstrated in the paper [Understanding R1-Zero-Like Training: A Critical Perspective](https://huggingface.co/papers/2503.14476) that the initial GRPO formulation introduces a response length bias. They show that while the DAPO formulation reduces this bias, it does not eliminate it completely. To fully remove this bias, they propose dividing by a constant instead of the sequence length, resulting in the following formulation:
+
+$$
+\mathcal{L}_{\text{Dr. GRPO}}(\theta) = - \frac{1}{LG} \sum_{i=1}^G \sum_{t=1}^{|o_i|} l_{i,t},
+$$
+
+This constant is recommended to be the maximum completion length. To use this formulation, set `loss_type="drgrpo"` in the [`GRPOConfig`].
```
Comment (lewtun) [inline docs/source/grpo_trainer.md:142]: turbo nit but I think we should follow the same convention we use in `DPOTrainer` to snake_case loss names:
```suggestion
This constant is recommended to be the maximum completion length. To use this formulation, set `loss_type="dr_grpo"` in the [`GRPOConfig`].
```

File: trl/trainer/grpo_config.py [hunk 3]
```diff
@@ -112,9 +112,19 @@ class GRPOConfig(TrainingArguments):
         scale_rewards (`bool`, *optional*, defaults to `True`):
             Whether to scale the rewards by dividing them by their standard deviation. If `True` (default), the rewards
             are normalized by the standard deviation, ensuring they have unit variance. If `False`, no scaling is
-            applied. The [Dr. GRPO](https://github.com/sail-sg/understand-r1-zero/blob/main/understand-r1-zero.pdf)
-            paper recommends not scaling the rewards, as scaling by the standard deviation introduces a question-level
-            difficulty bias.
+            applied. The [Dr. GRPO paper](https://huggingface.co/papers/2503.14476) recommends not scaling the rewards,
+            as scaling by the standard deviation introduces a question-level difficulty bias.
+        loss_type (`str`, *optional*, defaults to `"bnpo"`):
```
Comment (lewtun) [inline trl/trainer/grpo_config.py:117]: What is `bnpo`? Would be good to have a reference to where it's defined (I thought we had DAPO as the default loss)

File: trl/trainer/grpo_config.py [hunk 4]
```diff
@@ -112,9 +112,19 @@ class GRPOConfig(TrainingArguments):
...
+            Specifies the loss formulation to use. Supported values are:
+
+            - `"bnpo"`: Token-level losses are aggregated by normalizing with the completion length
+                within the local batch. Note that normalization is performed over the local batch only, so results may
+                slightly vary depending on the local batch size, despite a constant effective batch size.
+            - `"drgrpo"`: Token-level losses are aggregated by normalizing with a global constant. This method was
+                introduced in the [Dr. GRPO paper](https://huggingface.co/papers/2503.14476) to eliminate length bias.
+                The value of the constant corresponds to `max_completion_length`.
```
Comment (lewtun) [inline trl/trainer/grpo_config.py:125]: If I understand correctly, @edbeeching was trying something slightly different in https://github.com/huggingface/trl/pull/3231 that did local scaling per batch instead of a global constant. Do you know if there's much difference between the two?

**Review by lewtun — APPROVED** [review]
LGTM with some nits and a question about what BNPO refers to

File: trl/trainer/grpo_config.py (hunk 4 above)
Comment (edbeeching) [reply trl/trainer/grpo_config.py:125]: They are roughly equivalent, I have closed my PR in favor of this one.

**Comment by qgallouedec** [comment]
Results mostly match, expect for the the loss and the grad norm, Dr GRPO seems to reduce the range, as expected.

Screenshot 2025-04-08 at 06 19 52
Screenshot 2025-04-08 at 06 19 33

File: trl/trainer/grpo_config.py (hunk 3 above)
Comment (qgallouedec) [reply trl/trainer/grpo_config.py:117]: In fact, I realized while doing this PR that it wasn't exactly DAPO that was being used, but a variant of BNPO as defined here :

Screenshot 2025-04-08 at 06 27 50

Let me try to clarify here. Losses per token are normalized by 
- GRPO: the length of the sequence
- DAPO: the average sequence length in the group
- BNPO: the average sequence length in the batch
- TRL: the average sequence length in the local batch*; this is what I call bnpo here, but it's not 100% correct
- Dr GRPO: by the maximum possible length of the completion 
- 
- * a batch is made up of `num_devices * gradient_accumulations` local batches

Comment (qgallouedec) [reply trl/trainer/grpo_config.py:117]: Special cases:
When
- `per_device_batch_size==num_generations`, TRL's BNPO is equivalent to DAPO
- `per_device_batch_size==1`, TRL's BNPO is equivalent to GRPO
- `gradient_accumualtion_steps==1` and `num_devices=1`,  TRL's BNPO is equivalent to the actual BNPO.

Comment (lkevinzc) [reply trl/trainer/grpo_config.py:121]: @qgallouedec  Thanks for the comprehensive support! A minor comment for your future consideration: Dr. GRPO does not constrain the constant normalizer to be MAX_LEN (although it's easier to just use that). This can affect the update scale (related to your recent tweet https://x.com/QGallouedec/status/1908741708021457357). In fact, different constant of `x` in the setting in your tweet can be absorbed into the constant normalizer we propose in the paper, and MAX_LEN is a convenient example.
